image_creator/image.py

```diff
@@ -347,6 +336,76 @@ def part_set_bootable(partnum, bootable):
             """Sets the bootable flag for a partition"""
             self.g.part_set_bootable(self.guestfs_device, partnum, bootable)
 
+        last_part = self._last_partition()
+        sector_size = self.g.blockdev_getss(self.guestfs_device)
+        start = last_part['part_start'] / sector_size
+
+        if is_logical(last_part):
+            partitions = self.g.part_list(self.guestfs_device)
+
+            logical = []  # logical partitions
+            for partition in partitions:
+                if partition['part_num'] < 4:
+                    continue
+                logical.append({
+                    'num': partition['part_num'],
+                    'start': partition['part_start'] / sector_size,
+                    'end': partition['part_end'] / sector_size,
+                    'id': part_get_id(partition['part_num']),
+                    'bootable': part_get_bootable(partition['part_num'])
+                })
+
+            logical[-1]['end'] = end  # new end after resize
+
+            # Recreate the extended partition
+            extended = filter(is_extended, partitions)[0]
+            part_del(extended['part_num'])
+            part_add('e', extended['part_start'] / sector_size, end)
+
+            # Create all the logical partitions back
+            for l in logical:
+                part_add('l', l['start'], l['end'])
+                part_set_id(l['num'], l['id'])
+                part_set_bootable(l['num'], l['bootable'])
+        else:
+            # Recreate the last partition
+            if self.meta['PARTITION_TABLE'] == 'msdos':
+                last_part['id'] = part_get_id(last_part['part_num'])
+
+            last_part['bootable'] = part_get_bootable(last_part['part_num'])
+            part_del(last_part['part_num'])
+            part_add('p', start, end)
+            part_set_bootable(last_part['part_num'], last_part['bootable'])
+
+            if self.meta['PARTITION_TABLE'] == 'msdos':
+                part_set_id(last_part['part_num'], last_part['id'])
+
+    def shrink(self, silent=False):
+        """Shrink the image.
+
+        This is accomplished by shrinking the last file system of the
+        image and then updating the partition table. The shrinked device is
+        returned.
+
+        ATTENTION: make sure umount is called before shrink
+        """
+        def get_fstype(partition):
+            """Get file system type"""
+            device = "%s%d" % (self.guestfs_device, partition['part_num'])
+            return self.g.vfs_type(device)
+
+        def is_extended(partition):
+            """Returns True if the partition is an extended partition"""
+            if self.meta['PARTITION_TABLE'] == 'msdos':
```

Comment:
Actually yes. This is different. In the other 2 cases, the `is_extended()` is executed inside a `if is_logical():` block. Logical and extended partitions are MSDOS partitions. The fact that the code is inside the block implies that it an MSDOS partition table. (I know, it's ugly. I probably need to clean it up).